block-producer/src/block_builder/prover/mod.rs

```diff
@@ -106,6 +109,38 @@ proc.compute_note_root
     # => [ROOT_{n-1}]
 end
 
+#! Stack: [num_produced_nullifiers, OLD_NULLIFIER_ROOT, NULLIFIER_VALUE,
+#!         NULLIFIER_0, ..., NULLIFIER_n]
+#! Output: [NULLIFIER_ROOT]
+proc.compute_nullifier_root
+    # assess if we should loop
+    dup neq.0
+    #=> [0 or 1, num_produced_nullifiers, OLD_NULLIFIER_ROOT, NULLIFIER_VALUE, NULLIFIER_0, ..., NULLIFIER_n ]
+    
+    while.true
+        #=> [num_nullifiers_left_to_update, ROOT_i, NULLIFIER_VALUE, NULLIFIER_i, ... ]
+
+        # Prepare stack for `smt::set`
+        movdn.12 movupw.2 dupw.2
+        #=> [NULLIFIER_VALUE, NULLIFIER_i, ROOT_i, NULLIFIER_VALUE, num_nullifiers_left_to_update, ... ]
+
+        exec.smt::set
+        #=> [OLD_VALUE, ROOT_{i+1}, NULLIFIER_VALUE, num_nullifiers_left_to_update, ... ]
+
+        # Check that OLD_VALUE == 0 (i.e. that nullifier was indeed not previously produced)
+        padw assert_eqw
+        #=> [ROOT_{i+1}, NULLIFIER_VALUE, num_nullifiers_left_to_update, ... ]
```

Comment:
small optimization observation: here we just want to assert that top 4 elements of the stack are zeros, right? If so, doing `assertz assertz assertz assertz` would be a bit faster (8 cycles vs. 19).

Reply:
yup! Fixed